### PR TITLE
Feat: change tax category NMA-1269

### PR DESCRIPTION
**Pull request by HashEngineering**
## How Has This Been Tested?

- [ ] QA (Mobile Team)

## Checklist:

- [ ] I have performed a self-review of my own code and added comments where necessary
- [ ] I have added or updated relevant unit/integration/functional/e2e tests

### CODE REVIEW

File: wallet/test/de/schildbach/wallet/transactions/TransactionExportTest.kt [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+            listOf(
+                CSVColumn("Empty 1") { _, _ -> "" },
+                CSVColumn("Empty 2", emptyColumn)),
+            listOf(),
+            mapOf<Sha256Hash, TransactionMetadata>()
+        )*/
+
+        //exporter.exportString()
+    }
+}
```
Comment (Syn-McJ) [inline wallet/test/de/schildbach/wallet/transactions/TransactionExportTest.kt:48]: This is a strange looking test with too many licenses :)

**Review by Syn-McJ — COMMENTED** [review]
@HashEngineering looks good but I have some comments on the usage of coroutines and flows. I suppose none of them are critical but I think it'd be better to fix some of those issues to avoid potential problems in the future.

File: common/src/main/java/org/dash/wallet/common/transactions/TransactionMetadata.kt [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ */
+package org.dash.wallet.common.transactions
+
+import androidx.room.Entity
+import androidx.room.Ignore
+import androidx.room.PrimaryKey
+import org.bitcoinj.core.Coin
+import org.bitcoinj.core.Sha256Hash
+
+@Entity(tableName = "transaction_metadata")
```
[the remark this answers is not in the document]
Comment (HashEngineering) [reply common/src/main/java/org/dash/wallet/common/transactions/TransactionMetadata.kt:24]: That is a good point. I figured that primary keys were made into indexes by default, so I will check that.  The timestamp may could also use an index.

Though these tables will be small for the most part (not 100K rows), but some users will have thousands of transactions.

File: wallet/src/de/schildbach/wallet/data/TransactionMetadataDao.kt [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import org.dash.wallet.common.transactions.TaxCategory
+import org.dash.wallet.common.transactions.TransactionMetadata
+
+/**
+ * @author Eric Britten
+ */
+@Dao
+interface TransactionMetadataDao {
+    @Insert(onConflict = OnConflictStrategy.REPLACE)
+    fun insert(transactionMetadata: TransactionMetadata)
```
[the remark this answers is not in the document]
Comment (HashEngineering) [reply wallet/src/de/schildbach/wallet/data/TransactionMetadataDao.kt:32]: Good point.

In BlockchainServiceImpl, there may be a need update this table, but we can make a non-suspended method for this purpose in the `TransactionMetadataService`.

File: wallet/src/de/schildbach/wallet/data/TransactionMetadataDao.kt [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+    @Query("SELECT * FROM transaction_metadata")
+    fun load(): LiveData<List<TransactionMetadata>>
```
[the remark this answers is not in the document]
Comment (HashEngineering) [reply wallet/src/de/schildbach/wallet/data/TransactionMetadataDao.kt:35]: As you stated earlier, making all methods `suspend` is probably the way to go.

File: common/src/main/java/org/dash/wallet/common/transactions/TransactionMetadata.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (Syn-McJ) [reply common/src/main/java/org/dash/wallet/common/transactions/TransactionMetadata.kt:24]: Yeah this must be true about the primary key

File: wallet/src/de/schildbach/wallet/di/DataProviderModule.kt [hunk 5]
```diff
@@ -47,4 +50,9 @@ object DataProviderModule {
     @Singleton
     @Provides
     fun provideDataSyncStatus(): DataSyncStatusService = ExploreDataSyncStatus()
+
+    @Singleton
+    @Provides
+    fun provideTransactionMetadata(transactionMetadataDao: TransactionMetadataDao,
```
[the remark this answers is not in the document]
Comment (HashEngineering) [reply wallet/src/de/schildbach/wallet/di/DataProviderModule.kt:56]: I can try this out.

File: wallet/src/de/schildbach/wallet/ui/TransactionResultViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        monitorTransactionMetadata()
+    }
+
+    /**
+     * This method uses a Job to keep track of an observer.  If [currentJob] exists, it will
+     * be canceled and then a new Job is created.
+     *
+     * This is necessary when this ViewModel is own by an activity that is creating dialogs
+     * that will be using the same ViewModel instance.  Otherwise, many observers will exist
+     * for the previous instances of the dialog.
```
[the remark this answers is not in the document]
Comment (HashEngineering) [reply wallet/src/de/schildbach/wallet/ui/TransactionResultViewModel.kt:58]: Originally I had tried `by viewModel()` but had a crash, but was missing `@AndroidEntryPoint` on the dialog fragment class.  The first option will be used.

File: wallet/src/de/schildbach/wallet/service/WalletTransactionMetadataService.kt [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    override suspend fun getTransactionMetadata(txid: Sha256Hash): TransactionMetadata? {
+        val transactionMetadata = transactionMetadataDao.loadSync(txid)
+        if (transactionMetadata == null) {
+            insertTransactionMetadata(txid)
+        }
+        return transactionMetadataDao.loadSync(txid)
+    }
+
+    @OptIn(ExperimentalCoroutinesApi::class)
+    override suspend fun observeTransactionMetadata(txid: Sha256Hash): Flow<TransactionMetadata?> {
```
[the remark this answers is not in the document]
Comment (HashEngineering) [reply wallet/src/de/schildbach/wallet/service/WalletTransactionMetadataService.kt:112]: Good idea.  This is what will be done.

File: wallet/src/de/schildbach/wallet/data/TransactionMetadataDao.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (HashEngineering) [reply wallet/src/de/schildbach/wallet/data/TransactionMetadataDao.kt:35]: Except for the methods that return a `Flow`

File: wallet/src/de/schildbach/wallet/di/DataProviderModule.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (HashEngineering) [reply wallet/src/de/schildbach/wallet/di/DataProviderModule.kt]: This idea works.

File: wallet/src/de/schildbach/wallet/ui/ToolsActivity.kt [hunk 8]
```diff
@@ -90,9 +89,16 @@ class ToolsActivity : BaseMenuActivity() {
                     getString(R.string.close))
                 dialog.show(supportFragmentManager, "requireSyncing")
             } else {
-                alertDialog = ExportTransactionHistoryDialogBuilder.createExportTransactionDialog(this,
-                    WalletApplication.getInstance()).buildAlertDialog()
-                alertDialog.show()
+                viewModel.getTransactionExporter()
+                viewModel.transactionExporter.observe(this) {
```
[the remark this answers is not in the document]
Comment (HashEngineering) [reply wallet/src/de/schildbach/wallet/ui/ToolsActivity.kt:93]: Right, the `transactionExporter` is never updated.

I wasn't sure how to send `transactionExporter` to the `ToolsActivity` to create the CSVExport Dialog.

**Review by Syn-McJ — APPROVED** [review]
Looks great, although the test is failing, please check.

**Comment by HashEngineering** [comment]
The tests pass locally on my system, I will restart the build.

**Comment by HashEngineering** [comment]
@Syn-McJ The failing test was moved to `androidTest` because Wallet requires a `Context` which requires the BLS shared libraries and I wasn't sure if `tests` could provide that support or if I could mock those requirements.  If there is a way to mock `Context` then I can do that in the next story.

**Comment by Syn-McJ** [comment]
@HashEngineering yes I had this problem with `SendCoinsTaskRunner`. I don't think I solved it properly, just ended up testing another method without `propagate`.
But with mockk (used in the wallet module tests) you should be able to do something like this:
```
mockkStatic(Context::class)
every { Context.propagate(any()) } just runs
```

**Pull request merged**